File: PixelDungeon/src/main/java/com/nyrds/retrodungeon/mobs/necropolis/Lich.java
package com.nyrds.retrodungeon.mobs.necropolis;

import com.nyrds.Packable;
import com.nyrds.retrodungeon.items.necropolis.BlackSkull;
import com.nyrds.retrodungeon.items.necropolis.BlackSkullOfMastery;
import com.watabou.noosa.Game;
import com.watabou.noosa.audio.Sample;
import com.watabou.pixeldungeon.Assets;
import com.watabou.pixeldungeon.Badges;
import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.actors.Actor;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.blobs.Blob;
import com.watabou.pixeldungeon.actors.blobs.ToxicGas;
import com.watabou.pixeldungeon.actors.buffs.Amok;
import com.watabou.pixeldungeon.actors.buffs.Blindness;
import com.watabou.pixeldungeon.actors.buffs.Paralysis;
import com.watabou.pixeldungeon.actors.buffs.Sleep;
import com.watabou.pixeldungeon.actors.buffs.Terror;
import com.watabou.pixeldungeon.actors.hero.HeroClass;
import com.watabou.pixeldungeon.actors.mobs.Boss;
import com.watabou.pixeldungeon.actors.mobs.Mob;
import com.watabou.pixeldungeon.actors.mobs.Skeleton;
import com.watabou.pixeldungeon.effects.CellEmitter;
import com.watabou.pixeldungeon.effects.Pushing;
import com.watabou.pixeldungeon.effects.particles.ShadowParticle;
import com.watabou.pixeldungeon.items.keys.SkeletonKey;
import com.watabou.pixeldungeon.items.potions.PotionOfHealing;
import com.watabou.pixeldungeon.items.wands.WandOfBlink;
import com.watabou.pixeldungeon.items.weapon.enchantments.Death;
import com.watabou.pixeldungeon.levels.Level;
import com.watabou.pixeldungeon.levels.Terrain;
import com.watabou.pixeldungeon.mechanics.Ballistica;
import com.watabou.pixeldungeon.scenes.GameScene;
import com.watabou.utils.Random;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;

/**
 * Created by DeadDie on 12.02.2016
 */
public class Lich extends Boss {

    private static final int SKULLS_BY_DEFAULT	= 3;
    private static final int SKULLS_MAX	= 4;
    private static final int HEALTH	= 200;
    private static final int SKULL_DELAY = 5;

    private RunicSkull activatedSkull;

    @Packable
    private boolean skullsSpawned = false;

	private HashSet<RunicSkull> skulls = new HashSet<>();

    public Lich() {
        hp(ht(HEALTH));
        exp = 25;
        defenseSkill = 23;

        IMMUNITIES.add( Paralysis.class );
        IMMUNITIES.add( ToxicGas.class );
        IMMUNITIES.add( Terror.class );
        IMMUNITIES.add( Death.class );
        IMMUNITIES.add( Amok.class );
        IMMUNITIES.add( Blindness.class );
        IMMUNITIES.add( Sleep.class );
    }

    private int timeToSkull = SKULL_DELAY;

    @Override
    protected boolean getCloser( int target ) {
        if (Dungeon.level.fieldOfView[target]) {
            jump();
            return true;
        } else {
            return super.getCloser( target );
        }
    }

    @Override
    protected boolean canAttack( Char enemy ) {
        return Dungeon.level.distance(getPos(), enemy.getPos()) < 4 && Ballistica.cast(getPos(), enemy.getPos(), false, true) == enemy.getPos();
    }

    @Override
    protected boolean doAttack( Char enemy ) {
        if (Dungeon.level.distance(getPos(), enemy.getPos()) <= 1) {
            return super.doAttack(enemy);
        } else {

            getSprite().zap(enemy.getPos());

            spend(1);

            if (hit(this, enemy, true)) {
                enemy.damage(damageRoll(), this);
            }
            return true;
        }
    }

    private void jump() {
        int newPos;

        for (int i = 0; i < 15; i++){

            newPos = Random.Int( Dungeon.level.getLength() );

            if(Dungeon.level.fieldOfView[newPos] &&
                    Dungeon.level.passable[newPos] &&
                    !Dungeon.level.adjacent( newPos, getEnemy().getPos()) &&
                    Actor.findChar( newPos ) == null)
            {
                getSprite().move( getPos(), newPos );
                move( newPos );
                spend( 1 / speed() );
                break;
            }
        }
    }

    //Runic skulls handling
    //***

    private void activateRandomSkull(){
       if(!skullsSpawned) {
           skullsSpawned = true;
           spawnSkulls();
    }

        if (!skulls.isEmpty()){
            if (activatedSkull != null){
                activatedSkull.Deactivate();
            }

            RunicSkull skull = getRandomSkull();
            if(skull == null){
                activatedSkull = null;
            } else{
                skull.Activate();
                activatedSkull = skull;
            }
        }
    }

    private RunicSkull getRandomSkull() {
        while(!skulls.isEmpty()){
            RunicSkull skull = Random.element(skulls);
            if(skull.isAlive()){
                return skull;
            }
            else{
                skulls.remove(skull);
            }
        }
        return null;
    }

    private void useSkull(){
        getSprite().zap(getPos(), null);

        switch (activatedSkull.getKind()) {
            case RunicSkull.RED_SKULL:
                PotionOfHealing.heal(this,0.07f * skulls.size());
                break;

            case RunicSkull.BLUE_SKULL:
                int i = 0;
                while (i < skulls.size()){
                    int pos = Dungeon.level.getEmptyCellNextTo(getPos());
                    if (Dungeon.level.cellValid(pos)) {
                        Skeleton skeleton = new Skeleton();
                        skeleton.setPos(pos);
                        Dungeon.level.spawnMob(skeleton, 0);
                        Actor.addDelayed(new Pushing(skeleton, getPos(), skeleton.getPos()), -1);
                        i++;
                    } else {
                        break;
                    }
                }
                break;

            case RunicSkull.GREEN_SKULL:
                GameScene.add( Blob.seed( getPos(), 30 * skulls.size(), ToxicGas.class ) );
                break;
        }
    }

    //***

    @Override
    protected boolean act() {
        timeToSkull--;
        if (timeToSkull < 0){
            timeToSkull = SKULL_DELAY;
            activateRandomSkull();
            if (activatedSkull != null) {
                useSkull();
            }
        }
        return super.act();
    }


    @Override
    public int damageRoll() {
        return Random.NormalIntRange( 12, 20 );
    }

    @Override
    public int defenseProc(Char enemy, int damage) {
        if (activatedSkull != null)
        {
            if(activatedSkull.getKind() == RunicSkull.PURPLE_SKULL){
                return 0;
            }
        }

        if (Random.Int(2) == 1 && this.isAlive()){
            jump();
        }

        return damage;
    }

    @Override
    public int attackSkill( Char target ) {
        return 30;
    }

    @Override
    public int dr() {
        return 15;
    }

    @Override
    public void die( Object cause ) {
        GameScene.bossSlain();
        if ( Dungeon.hero.heroClass == HeroClass.NECROMANCER){
            Dungeon.level.drop( new BlackSkullOfMastery(), getPos() ).sprite.drop();
        }
        else {
            Dungeon.level.drop( new BlackSkull(), getPos() ).sprite.drop();
        }
        super.die( cause );
        Dungeon.level.drop( new SkeletonKey(), getPos() ).sprite.drop();
        //Kill everything
        skulls.clear();
        Mob mob = Dungeon.level.getRandomMob();
        while(mob != null){
            mob.remove();
            mob = Dungeon.level.getRandomMob();
        }
        Badges.validateBossSlain(Badges.Badge.LICH_SLAIN);
    }

    private void spawnSkulls(){
        int nSkulls = SKULLS_BY_DEFAULT;
        if(Game.getDifficulty() == 0){
            nSkulls = 2;
        }
        else if(Game.getDifficulty() > 2){
            nSkulls = SKULLS_MAX;
        }

        Level level = Dungeon.level;
        ArrayList<Integer> pedestals = level.getAllTerrainCells(Terrain.PEDESTAL);

        Collections.shuffle(pedestals);

        Sample.INSTANCE.play( Assets.SND_CURSED );

        for (int i = 0;i < nSkulls && i < pedestals.size();++i) {
            RunicSkull skull = RunicSkull.makeNewSkull(i);
            level.spawnMob(skull);

            CellEmitter.center(pedestals.get(i)).burst( ShadowParticle.CURSE, 8 );
            WandOfBlink.appear(skull, pedestals.get(i));

            skulls.add(skull);
        }
    }
}

File: scripts/mobs/Lich.lua
--[[
  Lich (batch 17d-3, was mobs/necropolis/Lich.java).
  Necropolis boss: jumps instead of walking while the enemy is visible
  (getCloser hook), and may jump right before an attack after taking a hit
  (defenceProc sets timeToJump, doAttack consumes it). Every 5 turns cycles
  the runic skulls: once, up to 4 skulls appear on the level pedestals
  (2 on easy difficulty, 4 on hard) and a live one is activated - RED heals
  the lich, BLUE summons hunting skeletons, GREEN vents toxic gas, PURPLE
  nullifies damage while active (defenceProc). Skulls are pure json - the
  lich drives them and keys each skull's variant by its cell (they never
  move). Killing the lich removes every non-pet mob on the level.
]]
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"
local WandOfBlink = luajava.bindClass("com.watabou.pixeldungeon.items.wands.WandOfBlink")
local ShadowParticle = luajava.bindClass("com.watabou.pixeldungeon.effects.particles.ShadowParticle")

local SKULL_DELAY = 5

local RED, BLUE, GREEN, PURPLE = 0, 1, 2, 3

-- java Level.adjacent is a raw cell-diff check (rows can "wrap"); port it
-- verbatim instead of a proper chebyshev
local function adjacent(level, a, b)
    local diff = math.abs(a - b)
    local W = level:getWidth()
    return diff == 1 or diff == W or diff == W + 1 or diff == W - 1
end

local function skullsOnLevel()
    local res = {}
    for _, m in pairs(RPD.Dungeon.level:getMobs()) do
        if m:getEntityKind() == "RunicSkull" and m:isAlive() then
            res[#res + 1] = m
        end
    end
    return res
end

local function spawnSkulls(self, data)
    data.skullsSpawned = true
    data.variantByPos = data.variantByPos or {}

    local nSkulls = 3
    local difficulty = RPD.GameLoop:getDifficulty()
    if difficulty == 0 then
        nSkulls = 2
    elseif difficulty > 2 then
        nSkulls = 4
    end

    local pedestals = {}
    for _, obj in pairs(RPD.Dungeon.level:getLevelObjects()) do
        if obj:getEntityKind() == "pedestal" then
            pedestals[#pedestals + 1] = obj:getPos()
        end
    end

    RPD.playSound("snd_cursed")

    -- java shuffled the pedestals and handed makeNewSkull(i) to the i-th
    -- pick: variants ride the random spawn order
    for i = 0, nSkulls - 1 do
        if #pedestals == 0 then
            break
        end
        local pos = table.remove(pedestals, math.random(#pedestals))

        local skull = RPD.MobFactory:mobByName("RunicSkull")
        WandOfBlink:appear(skull, pos)
        data.variantByPos[pos] = i
        RPD.Sfx.CellEmitter:get(pos):burst(ShadowParticle.CURSE, 8)
    end
end

local function useSkull(self, data)
    self:getSprite():zap(self:getPos())

    local alive = #skullsOnLevel()
    local variant = data.variantByPos[data.activatedSkullPos]

    if variant == RED then
        -- java: PotionOfHealing.heal(this, 0.07f * skulls.size())
        self:heal(math.floor(self:ht() * 0.07 * alive), self)
        self:detachBuff("Poison")
        self:detachBuff("Cripple")
        self:detachBuff("Weakness")
        self:detachBuff("Bleeding")
    elseif variant == BLUE then
        for _ = 1, alive do
            local skeleton = RPD.CharUtils:spawnOnNextCell(self, "Skeleton", 999)
            if not skeleton:valid() then
                break
            end
            RPD.setAi(skeleton, "Hunting")
        end
    elseif variant == GREEN then
        RPD.placeBlob(RPD.Blobs.ToxicGas, self:getPos(), 30 * alive)
    end
end

local function jump(self, data)
    data.timeToJump = false

    local level = RPD.Dungeon.level
    local enemy = self:getEnemy()
    local enemyPos = enemy and enemy:getPos() or -1

    for _ = 1, 15 do
        local newPos = math.random(0, level:getLength() - 1)
        if level.fieldOfView[newPos + 1]
                and level.passable[newPos + 1]
                and not adjacent(level, self:getPos(), enemyPos)
                and RPD.Actor:findChar(newPos) == nil then

            self:move(newPos)
            self:spend(1 / self:speed())
            break
        end
    end
end

return mob.init{
    spawn = function(self, level)
        local heroClass = RPD.Dungeon.hero:getHeroClass():name()
        local bag = self:getBelongings()

        if not bag:getItem("SkeletonKey") then
            self:collect(RPD.item("SkeletonKey"))
        end

        if heroClass == "NECROMANCER" then
            if not bag:getItem("BlackSkullOfMastery") then
                self:collect(RPD.item("BlackSkullOfMastery"))
            end
        elseif not bag:getItem("BlackSkull") then
            self:collect(RPD.item("BlackSkull"))
        end
    end,

    act = function(self)
        local data = mob.restoreData(self)
        if data.timeToSkull == nil then
            data.timeToSkull = SKULL_DELAY
        end

        data.timeToSkull = data.timeToSkull - 1
        if data.timeToSkull < 0 then
            data.timeToSkull = SKULL_DELAY

            local skulls = skullsOnLevel()
            if #skulls == 0 and not data.skullsSpawned then
                spawnSkulls(self, data)
                skulls = skullsOnLevel()
            end

            if #skulls > 0 then
                data.activatedSkullPos = skulls[math.random(#skulls)]:getPos()
                useSkull(self, data)
            else
                data.activatedSkullPos = nil
            end
        end
    end,

    getCloser = function(self, target, ignorePets)
        local level = RPD.Dungeon.level
        if level.fieldOfView[target + 1] then
            -- refund doStepTo's pre-charge; jump() spends the real cost
            self:spend(-1 / self:speed())
            jump(self, mob.restoreData(self))
            return true
        end
        return nil
    end,

    doAttack = function(self, enemy)
        local data = mob.restoreData(self)
        if data.timeToJump then
            jump(self, data)
        end
        return nil
    end,

    defenceProc = function(self, enemy, damage)
        local data = mob.restoreData(self)
        if data.activatedSkullPos ~= nil
                and (data.variantByPos or {})[data.activatedSkullPos] == PURPLE then
            return 0
        end

        if math.random(2) == 1 then
            data.timeToJump = true
        end

        return damage
    end,

    die = function(self, cause)
        RPD.CharUtils:validateBossSlain("LICH_SLAIN")

        -- this hook runs before Char.die's destroy(): the lich is still in
        -- the level list, and remove() would quietly eat its own loot - so
        -- java's post-super.die wipe is ported with an explicit self skip
        for _, m in pairs(RPD.Dungeon.level:getMobs()) do
            if m ~= self and not m:isPet() then
                m:remove()
            end
        end
    end,
}
